src/cockpit/channels/filesystem.py

```diff
@@ -48,46 +55,167 @@ def tag_from_fd(fd):
         return None
 
 
-class FsListChannel(Channel):
-    payload = 'fslist1'
+class AttributeReporter:
+    uid_cache: Dict[int, Union[str, int]]
+    gid_cache: Dict[int, Union[str, int]]
+    getters: Dict[str, Callable[[os.stat_result], JsonDocument]]
+
+    def get_tag(self, buf: os.stat_result) -> str:
+        return tag_from_stat(buf)
 
-    def send_entry(self, event, entry):
-        if entry.is_symlink():
-            mode = 'link'
-        elif entry.is_file():
-            mode = 'file'
-        elif entry.is_dir():
-            mode = 'directory'
+    def get_type(self, buf: os.stat_result) -> str:
+        if stat.S_ISDIR(buf.st_mode):
+            return 'directory'
+        elif stat.S_ISREG(buf.st_mode):
+            return 'file'
+        elif stat.S_ISLNK(buf.st_mode):
+            return 'link'
         else:
-            mode = 'special'
+            return 'special'
+
+    def get_mode(self, buf: os.stat_result) -> int:
+        return buf.st_mode & 0o777
+
+    def get_owner(self, buf: os.stat_result) -> Union[int, str]:
+        if buf.st_uid not in self.uid_cache:
+            try:
+                self.uid_cache[buf.st_uid] = pwd.getpwuid(buf.st_uid).pw_name
+            except KeyError:
+                self.uid_cache[buf.st_uid] = buf.st_uid
+        return self.uid_cache[buf.st_uid]
+
+    def get_group(self, buf: os.stat_result) -> Union[int, str]:
+        if buf.st_gid not in self.gid_cache:
+            try:
+                self.gid_cache[buf.st_gid] = grp.getgrgid(buf.st_gid).gr_name
+            except KeyError:
+                self.gid_cache[buf.st_gid] = buf.st_gid
+        return self.gid_cache[buf.st_gid]
+
+    def get_size(self, buf: os.stat_result) -> int:
+        return buf.st_size
 
-        self.send_message(event=event, path=entry.name, type=mode)
+    def get_modified(self, buf: os.stat_result) -> float:
+        return buf.st_mtime
+
+    def report(self, path: str, dir_fd: int = -1) -> JsonObject:
+        buf = os.lstat(path, dir_fd=dir_fd)
+        return {attr: getter(buf) for attr, getter in self.getters.items()}
+
+    def __init__(self, attrs: Sequence[str]) -> None:
+        self.uid_cache = {}
+        self.gid_cache = {}
+        self.getters = {attr: getattr(self, f'get_{attr}') for attr in attrs}
+
+
+class WatchAndListCommon(Channel, PathListener):
+    children: ClassVar[bool]
+
+    reporter: AttributeReporter
+    watch: Optional[PathWatch] = None
+    fd: Optional[Handle] = None
+    exists: Optional[bool] = None
+    path: str
+
+    def event(self, event: str, name: str) -> None:
+        assert self.fd is not None
+
+        attrs: JsonObject = {'event': event, 'path': name}
+
+        if event in ('present', 'created', 'attribute-changed'):
+            try:
+                attrs.update(self.reporter.report(name, dir_fd=self.fd))
```

Comment:
This will not work when you use `fswatch1` as we have `self.watch and not `self.fd`